sygnal/apnspushkin.py

```diff
@@ -79,9 +79,18 @@ class ApnsPushkin(ConcurrencyLimitedPushkin):
     Relays notifications to the Apple Push Notification Service.
     """
 
-    # Errors for which the token should be rejected
-    TOKEN_ERROR_REASON = "Unregistered"
-    TOKEN_ERROR_CODE = 410
+    # Errors for which the token should be rejected and not reused
+    # See https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/handling_notification_responses_from_apns
+    # for the full list of possible errors.
+    TOKEN_ERRORS = {
+        (400, "BadDeviceToken"),  # Invalid token
+        # `DeviceTokenNotForTopic` may be due to a bad token or a Sygnal
+        # misconfiguration. It's alright for the homeserver to forget a valid token,
+        # since iOS clients are expected to re-upload their device tokens periodically.
```

Comment:
In particular, Element iOS re-uploads its push token every time it's foregrounded (along with syncing and all the other stuff it does).

https://github.com/matrix-org/matrix-ios-kit/blob/b85b736313bec0592bd1cabc68035d97f5331137/MatrixKit/Models/Account/MXKAccount.m#L1152

Reply:
> since iOS clients are expected to re-upload their device tokens periodically

Is this just part of how APNS works or is it specific to Element iOS?

Reply:
Bad wording on my part here. It's specific to Element iOS and anything else built on matrix-ios-kit.

At worst, users would need to log out and back in again once the config is fixed. Though if you're in control of the Sygnal instance, you're presumably also responsible for the associated app and can update it to re-upload tokens.